File: Amazing Graphs/forest_fire.cpp
/*
    C++
    forest_fire.cpp
    2019.08
    https://oeis.org/A229037
Sequence of positive integers where each is chosen 
to be as small as possible subject to the condition that 
no three terms a(j), a(j+k), a(j+2k) (for any j and k) 
form an arithmetic progression.
*/


#include <iostream>
#include <fstream>
#include <array>


const int NO_POINTS = 100;


bool three_points_same_distance(std::array<int, NO_POINTS> nums, int X) {
    for (int x=X-1; x>=X/2; --x) {
        if ((2*nums[x] - nums[X]) == nums[2*x - X]) {
            return true;
        }
    }
    return false;
}


std::array<int, NO_POINTS> forest_fire() {
    std::array<int, NO_POINTS> nums = {1,1};
    int attempt;
    for (int X=2; X<NO_POINTS; ++X) {
        std::cout << X << "\n";
        attempt = 1;
        while(1) {
            nums[X] = attempt;
            if (three_points_same_distance(nums, X)) {
                attempt += 1;
            } else {
                break;
            }
        }
    }
    return nums;
}


int main() {
    std::array<int, NO_POINTS> nums;
    nums = forest_fire();

    std::cout << "Writing to file \'test.txt\'" << "\n";
    std::ofstream my_file;

    my_file.open("test.txt");
    my_file << "[";
    for (int i=0; i<NO_POINTS-1; i++) {
        my_file << nums[i] << ", ";
    }
    my_file << nums[NO_POINTS-1] << "]" << "\n";
    my_file.close();
    
    return 0;
}

File: Amazing Graphs/forest_fire.py
#!/usr/bin/env python3

"""
    forest_fire.py
    2019.08
    https://oeis.org/A229037
Sequence of positive integers where each is chosen 
to be as small as possible subject to the condition that 
no three terms a(j), a(j+k), a(j+2k) (for any j and k) 
form an arithmetic progression.

It takes too long for a large number of points.
"""

import matplotlib.pyplot as plt


NO_POINTS = 2000


def three_points_same_distance(lst):
    X = len(lst) - 1
    Y = lst[-1]
    for idx, y in enumerate(lst[-2:len(lst)//2-1:-1], 2):
        x = len(lst) - idx
        if (2*y - Y) == lst[2*x - X]:
            return True
    return False


def forest_fire(size):
    nums = [1, 1]
    while len(nums) < size:
        attempt = 1
        while True:
            nums.append(attempt)
            if three_points_same_distance(nums):
                nums.pop()
                attempt += 1
            else:
                break
    return nums


burntforest = forest_fire(NO_POINTS)

fig, ax = plt.subplots(1)

ax.plot(burntforest, ".")
ax.minorticks_on()
ax.grid(True, which="both")
plt.show()
